Message to Scalar flexibility

BasileiosKal opened the pull request:
Fix #245

Define input messages as octet strings and explicitly call `messages_to_scalars` in each operation.

Note this makes #266 obsolete, since it enforces consistency on the input arguments (everything are octets now).

Based on #267 to avoid the extra explicit checks (would be couple of steps)

BasileiosKal commented:
Discussed on the WG call on the 19th of June, will define the Input of `messages_to_scalars` to be typed elements instead of octet strings.

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -586,8 +584,9 @@ Inputs:
                  supplied, it defaults to an empty string.
-- disclosed_messages (OPTIONAL), a vector of scalars. If not supplied,
-                                 it defaults to the empty array "()".
+- disclosed_messages (OPTIONAL), a vector of input\_messages. If not
```
```suggestion
- disclosed_messages (OPTIONAL), a vector of input_messages. If not
```

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
````diff
@@ -741,37 +740,69 @@ Procedure:
+(scalar_1, ..., scalar_L) = messages_to_scalars((msg_1, ..., msg_L))
+```
+
+Every `messages_to_scalars` operation MUST define a unique `MESSAGES_2_SCALARS_ID` value to be used by the ciphersuite. This value MUST only contain ASCII encoded characters with codes between 0x21 and 0x7e (inclusive) and MUST end with an underscore (ASCII code: 0x5f), other than the last character the string MUST not contain any other underscores (ASCII code: 0x5f).
````
```suggestion
Every `messages_to_scalars` operation MUST define a unique `MESSAGES_TO_SCALARS_ID` value to be used by the ciphersuite. This value MUST only contain ASCII encoded characters with codes between 0x21 and 0x7e (inclusive) and MUST end with an underscore (ASCII code: 0x5f), other than the last character the string MUST not contain any other underscores (ASCII code: 0x5f).
```

I think it's better to spell it out.

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
````diff
@@ -741,37 +740,69 @@ Procedure:
+7. return (msg_scalar_1, ..., msg_scalar_L)
+```
+
+The `MESSAGES_2_SCALARS_ID` of the `hash_messages_to_scalar` is defines as,
````
```suggestion
The `MESSAGES_TO_SCALARS_ID` of the `hash_messages_to_scalar` is defines as,
```

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
````diff
@@ -1197,11 +1229,13 @@ This section defines the format for a BBS ciphersuite. It also gives concrete ci
 
 ```
-  "BBS_" || H2C_SUITE_ID || ADD_INFO
+  "BBS_" || H2C_SUITE_ID || MESSAGES_2_SCALARS_ID || ADD_INFO
````
```suggestion
  "BBS_" || H2C_SUITE_ID || MESSAGES_TO_SCALARS_ID || ADD_INFO
```

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -1197,11 +1229,13 @@ This section defines the format for a BBS ciphersuite. It also gives concrete ci
 
   *  H2C\_SUITE\_ID is the suite ID of the hash-to-curve suite used to define the hash_to_curve function.
 
+  *  MESSAGES\_2\_SCALARS\_ID is the ID of the messages\_to\_scalars operation, as defined in (#messages-to-scalars).
```
```suggestion
  *  MESSAGES\_TO\_SCALARS\_ID is the ID of the messages\_to\_scalars operation, as defined in (#messages-to-scalars).
```

andrewwhitehead commented:
I put in some suggestions for using `MESSAGES_TO_SCALARS_ID`, but perhaps `OCTETS_TO_SCALARS_ID` would make more sense and avoid overuse of the term 'messages'?

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -136,15 +136,15 @@ R
+scalar
+: An integer between 0 and r-1, where r is the prime order of the selected groups, defined by each ciphersuite (see also [Notation](#notation)).
+
+input_message
+: An input message to be signed by the signature scheme. An input_message can either be either an octet string or a scalar.
```
```suggestion
input\_message
: An input message to be signed by the signature scheme. An input\_message can either be either an octet string or a scalar.
```

BasileiosKal commented:
Updated the operation. As discussed in the WG, added a `map_to_scalar` operation, that will map a `(msg, idx)` pair to a scalar.

1. Turns out that if a msg is a scalar, simply returning it (while hashing the octet strings) is not secure. There is no way to know that what was signed is the octet string `msg` and not the `scalar = hash_to_scalar(msg)`.
2. Did not use `octet_to_scalar` for the same reason.

cc @andrewwhitehead i think the PR is ready for re-review!

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -765,38 +764,104 @@ When defining a new `create_generators` procedure, the most important property i
+
+To define different ways with which messages can be mapped to scalars, an application can define a new `map_to_scalar` operation, as part of a new ciphersuite. A new `map_to_scalar` function is REQUIRED to adhere to the following security rules:
+
+1. It MUST return unique values for different `msg` inputs.
```
Within reasonable guarantees of collision avoidance?

BasileiosKal replied:
Updated! Thanks!

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -765,38 +764,104 @@ When defining a new `create_generators` procedure, the most important property i
+#### Map to Scalar as Hash
 
-This operation takes an input message and maps it to a scalar value via a cryptographic hash function for the given curve. The operation takes also as an optional input a domain separation tag (dst). If a dst is not supplied, its value MUST default to the octet string returned from ciphersuite\_id || "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_", where ciphersuite\_id is the ASCII string representing the un ...
+This document specifies the following `map_to_scalar` operation, called `map_to_scalar_as_hash`, that uses `hash_to_scalar` as defined in (#hash-to-scalar). Although for extendability reasons, the `map_to_scalar` operation accepts messages that can be either an octet string or a scalar value (as to support protocol specific preprocessing of a message), the `map_to_scalar_as_hash` operation used b ...
```
Hm, this seems to mean that any application which wants to pre-process the messages as mentioned above, must use a ciphersuite not defined by this document, with a different mapping function.

BasileiosKal replied:
Yeap that was the indent. It seems that there is no way to define protocol-agnostic handling of scalar values (i.e., something like `if msg a scalar, return msg` does not work). 

Rules need to be more strict and also consider the index of a message (for example the 1st message is expected to be a scalar, etc). So a ciphersuite defined protocol-specific mapping seems unavoidable.

BasileiosKal replied:
You can always also define the `map_to_scalar` operation to just accept scalars and be the identity, and then apply whatever pre processing you want??

christianpaquin commented:
As mentioned on the discussion in today's WG call: U-Prove allows both direct encoding and hashing the attributes. This is useful, within _one_ ciphersuite, to allow encoding attributes of different types in one credential (e.g.: hash name, address, etc., but encode directly age and expiration to allow range proofs). See the `e` values in the Issuer params (section 2.3.1 of the [crypto spec](https://github.com/microsoft/uprove-node-reference/blob/main/doc/U-Prove%20Cryptographic%20Specification%20V1.1%20Revision%205.pdf)).

BasileiosKal commented:
Discussed on the WG call on the 10th of July. Merging and will open an Issue to keep track of the conversation.
